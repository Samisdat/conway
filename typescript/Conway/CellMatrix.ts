import {Cell} from './cell';
import {Position} from './position';

interface CellMatrixInterface {
    [index: string]: Cell;
}

export class CellMatrix {

    private matrix: CellMatrixInterface = {};

    public get(position: Position): Cell | undefined {

        if (undefined !== this.matrix[position.toString()]) {
            return this.matrix[position.toString()];
        }

        return undefined;


    }

    public add(cell: Cell): void {

        if (undefined !== this.matrix[cell.position.toString()]) {
            console.log('argh');
        }

        this.matrix[cell.position.toString()] = cell;

    }

    public remove(cell: Cell): void {

        if (undefined !== this.matrix[cell.position.toString()]) {
            delete this.matrix[cell.position.toString()];
        }

    }

    public all(): Cell[] {

        const cells: Cell[] = [];

        for (const index in this.matrix) {

            cells.push(this.matrix[index]);

        }

        return cells;

    }


}
