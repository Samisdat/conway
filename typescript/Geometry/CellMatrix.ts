import {Position} from '@Conway/Geometry/Position';

interface CellMatrixInterface {
    [index: string]: Position;
}

interface Dimension {
    width: number;
    height: number;
}

export class CellMatrix {

    private matrix: CellMatrixInterface = {};

    public has(position: Position): boolean {

        return (undefined !== this.matrix[position.toString()]);

    }

    public add(position: Position): void {

        if (true === this.has(position)) {
            return;
        }

        this.matrix[position.toString()] = position;

    }

    public remove(position: Position): void {

        if (false === this.has(position)) {
            return;
        }

        delete this.matrix[position.toString()];

    }

    public all(): Position[] {

        const positions: Position[] = [];

        for (const index in this.matrix) {

            positions.push(this.matrix[index]);

        }

        return positions;

    }

    private getDimension(): Dimension {

        const dimension: Dimension = {
            width: 0,
            height: 0
        }

        let left = 0;
        let right = 0;

        let top = 0;
        let bottom = 0;

        for (const position of this.all()) {

            if (position.x < left) {
                left = position.x;
            }

            if (position.x > right) {
                right = position.x;
            }

            if (position.y < bottom) {
                bottom = position.y;
            }

            if (position.y > top) {
                top = position.y;
            }

        }

        if (0 > left && 0 < right) {
            dimension.width = right - left;
        }
        else{
            dimension.width = right - left;
        }

        if (0 > bottom && 0 < top) {
            dimension.height = top - bottom;
        }
        else{
            dimension.height = top - bottom;
        }

        return dimension;

    }

    public width(): number {

        return this.getDimension().width;

    }

    public height(): number {

        return this.getDimension().height;

    }

}
