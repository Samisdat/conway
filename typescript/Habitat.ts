import {Position} from '@Conway/Geometry/Position';
import {Population} from '@Conway/Population';
import {GridCell} from '@Conway/Frontend/Grid/GridCell';
import {CellTypesFactory} from '@Conway/Frontend/CellType/CellTypesFactory';
import {GridDimension} from '@Conway/Frontend/Grid/GridDimension';

interface MatrixCol {
    [index: number]: number;
}

interface MatrixRow {
    [index: number]: MatrixCol;
}

export class Habitat {

    private gridDimension: GridDimension;
    private sourcePosition: Position;

    private offset: Position;

    private cells: GridCell[] = [];

    public readonly center: Position;

    private absoluteMap: MatrixRow = {};

    constructor(
        population: Population,
        gridDimension: GridDimension,
        sourcePosition: Position,
        offset: Position
    ) {

        Habitat.ensureOffsetBetweenMinusOneAndOne(offset);

        this.gridDimension = gridDimension;

        this.sourcePosition = sourcePosition;

        this.offset = offset;

        let relativePointer = new Position(0, 0);
        let absolutePointer = this.sourcePosition as Position;

        const inMiddleOfGrid = new Position(
            Math.floor(this.gridDimension.rows / 2),
            Math.floor(this.gridDimension.cols / 2)
        );


        this.center = relativePointer.move(inMiddleOfGrid);

        relativePointer = relativePointer.move(inMiddleOfGrid.inverse());

        absolutePointer = absolutePointer.move(inMiddleOfGrid.inverse());

        this.createGridCells(relativePointer, absolutePointer);

        this.markLivingCells(population);

    }

    private createGridCells(relativePointer: Position, absolutePointer: Position){

        const moveRight = new Position(
            1,
            0
        );

        // @TODO Use Bounds

        for (let top = 0; top < this.gridDimension.cols; top += 1) {

            for (let right = 0; right < this.gridDimension.rows; right += 1) {

                const gridCell = new GridCell(
                    relativePointer,
                    absolutePointer,
                    this.offset
                );

                const absolutePosition = gridCell.absolutePosition;

                if (undefined === this.absoluteMap[absolutePosition.x]) {
                    this.absoluteMap[absolutePosition.x] = {};
                }

                this.absoluteMap[absolutePosition.x][absolutePosition.y] = this.cells.length;

                this.cells.push(gridCell);

                relativePointer = relativePointer.move(
                    moveRight
                );

                absolutePointer = absolutePointer.move(moveRight);

            }

            relativePointer = relativePointer.move(
                new Position(-1 * this.gridDimension.rows, 1),
            );

            absolutePointer = absolutePointer.move(
                new Position(-1 * this.gridDimension.rows, 1),
            );

        }

    }

    private markLivingCells(population: Population){

        for (const positionWithLivingCells of population.get()) {

            if (undefined === this.absoluteMap[positionWithLivingCells.x]) {
                continue;
            }
            if (undefined === this.absoluteMap[positionWithLivingCells.x][positionWithLivingCells.y]) {
                continue;
            }

            const cellIndex = this.absoluteMap[positionWithLivingCells.x][positionWithLivingCells.y];

            this.cells[cellIndex].setType(CellTypesFactory.get().living());

        }

    }

    private static ensureOffsetBetweenMinusOneAndOne(offset: Position) {

        if (1 < Math.abs(offset.x )) {
            throw new Error('offset.x must be between -1 and 1');
        }

        if (1 < Math.abs(offset.y)) {
            throw new Error('offset.y must be between -1 and 1');
        }


    }

    public getSourcePosition(): Position {
        return this.sourcePosition;
    }

    public getOffset(): Position {
        return this.offset;
    }

    public getCols(): number {
        return this.gridDimension.cols;
    }

    public getRows(): number {
        return this.gridDimension.rows;
    }

    public getCells() {
        return this.cells;
    }

}
