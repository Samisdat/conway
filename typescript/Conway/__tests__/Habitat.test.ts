import {Position} from '../Position';
import {Habitat} from '../Habitat';
import {Patterns} from '../Patterns';

import serializer from '../../../jest-serialize-conway';
import {CellMatrix} from '../CellMatrix';

describe('Habitat', () => {

    let cellMatrix: CellMatrix;

    beforeEach(function () {

        expect.addSnapshotSerializer(serializer);

        cellMatrix = new CellMatrix();

    });

    test('can be created', () => {

        let habitat = new Habitat(
            cellMatrix,
            1000
        );

        expect(habitat).toBeInstanceOf(Habitat);

    });

    test('a not existing cell is not alive', () => {

        cellMatrix.add(
            new Position(0, 1)
        );

        let habitat = new Habitat(
            cellMatrix,
            1000
        );

        expect(habitat.getAllCells().length).toBe(1);

        habitat.elapse();

        expect(habitat.getAllCells().length).toBe(0);

    });


    test('snapshot', () => {

        const patterns = new Patterns();
        cellMatrix.seedPattern(patterns.get('blinker'));

        let habitat = new Habitat(
            cellMatrix,
            1000
        );

        expect(habitat.getMatrix()).toMatchSnapshot();

        habitat.elapse();

        expect(habitat.getMatrix()).toMatchSnapshot();

        habitat.elapse();

        expect(habitat.getMatrix()).toMatchSnapshot();
    });

    test('elapse with one cell', () => {

        cellMatrix.add(
            new Position(0, 1)
        );

        let habitat = new Habitat(
            cellMatrix,
            1000
        );

        expect(habitat.get()).toStrictEqual([new Position(0, 1)]);

    });

    test('elapse with a blinker', () => {

        cellMatrix.add(
            new Position(0, 0)
        );
        cellMatrix.add(
            new Position(0, 1)
        );
        cellMatrix.add(
            new Position(0, 2)
        );

        let habitat = new Habitat(
            cellMatrix,
            1000
        );

        expect(habitat.get()).toStrictEqual([
            new Position(0, 0),
            new Position(0, 1),
            new Position(0, 2)
        ]);

        habitat.elapse();

        expect(habitat.get()).toStrictEqual([
            new Position(0, 1),
            new Position(-1, 1),
            new Position(1, 1)
        ]);

        // check garbage collection
        expect(habitat.getAllCells().length).toBe(3);


        habitat.elapse();

        expect(habitat.get()).toStrictEqual([
            new Position(0, 1),
            new Position(0, 0),
            new Position(0, 2)
        ]);

        expect(habitat.getAllCells().length).toBe(3);

    });

});

jest.useFakeTimers();

describe('Habitat aging with interval', function() {

    let habitat: Habitat;

    beforeEach(() => {

        const patterns = new Patterns();

        const cellMatrix = new CellMatrix();

        cellMatrix.seedPattern(patterns.get('blinker'));

        habitat = new Habitat(
            cellMatrix,
            50
        );

    });


    test('should increase position', function() {

        expect(habitat.getMatrix()).toMatchSnapshot();

        habitat.startAging();

        jest.advanceTimersByTime(51);

        expect(habitat.getMatrix()).toMatchSnapshot();

    });

});
