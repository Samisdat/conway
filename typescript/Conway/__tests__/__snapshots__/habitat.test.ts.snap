// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Habitat snapshot: blinker-next 1`] = `
OOO
OXO
OXO
OXO
OOO
`;

exports[`Habitat snapshot: blinker-start 1`] = `
OOOOO
OXXXO
OOOOO
`;

exports[`Habitat snapshot: blinker-start 2`] = `
OOOOO
OXXXO
OOOOO
`;
